rviz_common/src/rviz_common/depth_cloud_mld.cpp

```
@@ -0,0 +1,574 @@
+        ++cloud_data_ptr;
+        *cloud_data_ptr = *reinterpret_cast<float *>(&color);
+        ++cloud_data_ptr;
+
+        ++point_count;
+
+        // if shadow point exists -> display it
+        if (depth < shadow_depth - shadow_distance_) {
+          // copy point data from shadow buffer to point cloud
+          memcpy(cloud_data_ptr, cloud_shadow_buffer_ptr, point_step);
+          cloud_data_ptr += 4;
+          ++point_count;
+        } else {
+          // save a copy of current point to shadow buffer
+          memcpy(cloud_shadow_buffer_ptr, cloud_data_pixel_ptr, point_step);
+
+          // reduce color intensity in shadow buffer
+          RGBA * color = reinterpret_cast<RGBA *>(cloud_shadow_buffer_ptr + sizeof(float) * 3);
+          color->red /= 2;
+          color->green /= 2;
+          color->blue /= 2;
+
+          // update shadow depth & time out
+          shadow_depth_[point_idx] = depth;
+          shadow_timestamp_[point_idx] = time_now;
+        }
+      } else {
+        // current depth pixel is invalid -> check shadow buffer
+        if (shadow_depth != 0) {
+          // copy shadow point to point cloud
+          memcpy(cloud_data_ptr, cloud_shadow_buffer_ptr, point_step);
+          cloud_data_ptr += 4;
+          ++point_count;
+        }
+      }
+
+      // increase color iterator pointer
+      if (color_img_ptr) {
+        ++color_img_ptr;
+      }
+    }
+  }
+
+  finalizingPointCloud(cloud_msg, point_count);
+
+  return cloud_msg;
+}
+
+
+template<typename T>
+void MultiLayerDepth::convertColor(
+  const sensor_msgs::msg::Image::ConstSharedPtr & color_msg,
+  std::vector<uint32_t> & rgba_color_raw)
+{
+  size_t i;
+  size_t num_pixel = color_msg->width * color_msg->height;
+
+  // query image properties
+  int num_channels = sensor_msgs::image_encodings::numChannels(color_msg->encoding);
+
+  bool rgb_encoding = false;
+  if (color_msg->encoding.find("rgb") != std::string::npos) {
+    rgb_encoding = true;
+  }
```

This check seems too simplistic.  There are a bunch of encodings, including RGB8, RGB16, YUYV, etc defined in https://github.com/ros2/common_interfaces/blob/rolling/sensor_msgs/include/sensor_msgs/image_encodings.hpp .  If we get an encoding that is not RGB or BGR, this code looks like it will assume it is RGB, which will probably end up with garbage.  Instead of that, I think we should check if the string starts with "rgb" or "bgr".  If it doesn't start with either of those, then we should probably return an error.

---

I will add a TODO